File: MiniDOMBase/DMD/Sistema/Attachments/CAttachmentChilds.vb
Imports minidom
Imports minidom.Sistema
Imports System.Xml.Serialization
Imports minidom.Anagrafica
Imports minidom.Databases
Imports minidom.Internals

Namespace Internals

    <Serializable>
    Public Class CAttachmentChilds
        Inherits CCollection(Of CAttachment)

        <NonSerialized> Private m_Parent As CAttachment

        Public Sub New()
            Me.m_Parent = Nothing
        End Sub


        Public Sub New(ByVal parent As CAttachment)
            Me.New()
            Me.Load(parent)
        End Sub

        Protected Overrides Sub OnInsert(index As Integer, value As Object)
            If (Me.m_Parent IsNot Nothing) Then DirectCast(value, CAttachment).SetParent(Me.m_Parent)
            MyBase.OnInsert(index, value)
        End Sub

        Protected Overrides Sub OnSet(index As Integer, oldValue As Object, newValue As Object)
            If (Me.m_Parent IsNot Nothing) Then DirectCast(newValue, CAttachment).SetParent(Me.m_Parent)
            MyBase.OnSet(index, oldValue, newValue)
        End Sub

        Protected Friend Sub SetParent(ByVal value As CAttachment)
            Me.m_Parent = value
            If (value IsNot Nothing) Then
                For Each a As CAttachment In Me
                    a.SetParent(value)
                Next
            End If
        End Sub


        Protected Sub Load(ByVal parent As CAttachment)
            Dim cursor As CAttachmentsCursor = Nothing

            If (parent Is Nothing) Then Throw New ArgumentNullException("parent")

            MyBase.Clear()

            Me.m_Parent = parent

            If (GetID(parent) = 0) Then Return

            Try
                cursor = New CAttachmentsCursor
                cursor.Stato.Value = ObjectStatus.OBJECT_VALID
                cursor.ParentID.Value = GetID(parent)
                cursor.IgnoreRights = True
                While Not cursor.EOF
                    Me.Add(cursor.Item)
                    cursor.MoveNext()
                End While

            Catch ex As Exception
                Sistema.Events.NotifyUnhandledException(ex)
                Throw
            Finally
                If (cursor IsNot Nothing) Then cursor.Dispose() : cursor = Nothing
            End Try
        End Sub

    End Class


End Namespace

File: MiniDOMBase/DMD/Sistema/Attachments/Childs/CAttachmentChilds.cs
using System;
using System.Collections;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using static minidom.Sistema;
using static minidom.Anagrafica;
using minidom.repositories;
using System.Collections.Generic;

namespace minidom.internals
{

    /// <summary>
    /// Collezione di allegati figlio
    /// </summary>
    [Serializable]
    public class CAttachmentChilds 
        : CCollection<Sistema.CAttachment>
    {
        [NonSerialized] private Sistema.CAttachment m_Parent;

        /// <summary>
        /// Costruttore
        /// </summary>
        public CAttachmentChilds()
        {
            m_Parent = null;
        }

        /// <summary>
        /// Costruttore
        /// </summary>
        /// <param name="parent"></param>
        public CAttachmentChilds(Sistema.CAttachment parent) : this()
        {
            Load(parent);
        }

        /// <summary>
        /// OnInsert
        /// </summary>
        /// <param name="index"></param>
        /// <param name="value"></param>
        protected override void OnInsert(int index, CAttachment value)
        {
            if (m_Parent is object)
                value.SetParent(m_Parent);

            base.OnInsert(index, value);
        }

        /// <summary>
        /// OnSet
        /// </summary>
        /// <param name="index"></param>
        /// <param name="oldValue"></param>
        /// <param name="newValue"></param>
        protected override void OnSet(int index, CAttachment oldValue, CAttachment newValue)
        {
            if (m_Parent is object)
                newValue.SetParent(m_Parent);

            base.OnSet(index, oldValue, newValue);
        }

        /// <summary>
        /// Imposta il genitore
        /// </summary>
        /// <param name="value"></param>
        protected internal void SetParent(Sistema.CAttachment value)
        {
            m_Parent = value;
            if (value is object)
            {
                foreach (Sistema.CAttachment a in this)
                    a.SetParent(value);
            }
        }

        /// <summary>
        /// Carica gli oggetti
        /// </summary>
        /// <param name="parent"></param>
        protected void Load(Sistema.CAttachment parent)
        {
            if (parent is null)
                throw new ArgumentNullException("parent");
            Clear();
            this.m_Parent = parent;
            if (DBUtils.GetID(parent, 0) == 0)
                return;

            using (var cursor = new Sistema.CAttachmentsCursor())
            { 
                cursor.Stato.Value = ObjectStatus.OBJECT_VALID;
                cursor.ParentID.Value = DBUtils.GetID(parent, 0);
                cursor.IgnoreRights = true;
                while (cursor.Read())
                {
                    Add(cursor.Item);
                }
            }              
        }
    }
}